src/Cawbird.vala

```diff
@@ -77,35 +97,86 @@ public class Cawbird : Adw.Application {
     Backend.Client client = new Backend.Client(Config.APPLICATION_ID,
                                                Config.PROJECT_NAME,
                                                "https://github.com/CodedOre/NewCaw",
-                                               "cawbird://authenticate");
+                                               "cawbird://authenticate",
+                                               state_path);
 
     // Load the previous program state
     this.hold ();
     client.load_state.begin ((obj, res) => {
       try {
         client.load_state.end (res);
+        Preferences.WindowManagement.load_state.begin (state_path, (obj, res) => {
+          try {
+            List<WindowAllocation?> stored_allocations = Preferences.WindowManagement.load_state.end (res);
+            foreach (WindowAllocation window_allocation in stored_allocations) {
+              window_allocations[window_allocation.session_id] = window_allocation;
+            }
 
-        // TODO: Reindroduce selected windows
-        if (client.sessions.get_n_items () > 0) {
-          foreach (Backend.Session session in client.sessions) {
-            var window = new MainWindow (this, session);
-            window.present ();
+            if (client.sessions.get_n_items () > 0) {
+              bool opened_window = false;
+              foreach (Backend.Session session in client.sessions) {
+                if (session.auto_start) {
+                  opened_window = true;
+                  var window = new MainWindow (this, session);
+                  window.present ();
+                }
+              }
+              if (!opened_window) {
+                // If nothing opened automatically, show the first session
+                var window = new MainWindow(this, client.sessions.get_item(0) as Backend.Session);
+                window.present();
+              }
+            } else {
+              var window = new MainWindow (this, null);
+              window.present ();
+            }
+          } catch (Error e) {
+            critical (@"Failed to load window state: $(e.message)");
+          } finally {
+            this.release ();
           }
-        } else {
-          var window = new MainWindow (this, null);
-          window.present ();
-        }
+        });
       } catch (Error e) {
         critical (@"Failed to load program state: $(e.message)");
-      } finally {
-        this.release ();
       }
     });
   }
 
+  internal void register_window (MainWindow window) {
+    string? session_id = window.session == null ? null : window.session.identifier;
+    if (session_id != null && window_allocations.contains (session_id)) {
+      WindowAllocation? window_allocation = window_allocations.get(session_id);
+      window.default_width = window_allocation.width;
+      window.default_height = window_allocation.height;
+    }
+    window.close_request.connect(update_window_allocation);
+    // TODO: Connect to session change
```

Comment:
Is this ToDo still relevant? The session seems to be updated on the close of the window.

Reply:
Good catch. That was my initial plans. But I don't think this approach needs it. It feels cleaner to save on close so that the user can switch through multiple profiles and not keep track of each one.